packages/core/src/transports/offline.ts

```diff
@@ -0,0 +1,169 @@
+
+export const MIN_DELAY = 100; // 100 ms
+export const START_DELAY = 5_000; // 5 seconds
+const MAX_DELAY = 3.6e6; // 1 hour
+
+function isReplayEnvelope(envelope: Envelope): boolean {
+  let isReplay = false;
+
+  forEachEnvelopeItem(envelope, (_, type) => {
+    if (type === 'replay_event') {
+      isReplay = true;
+    }
+  });
+
+  return isReplay;
+}
+
+function log(msg: string, error?: Error): void {
+  __DEBUG_BUILD__ && logger.info(`[Offline]: ${msg}`, error);
+}
+
+interface OfflineStore {
+  insert(env: Envelope): Promise<void>;
+  pop(): Promise<Envelope | undefined>;
+}
+
+export type CreateOfflineStore = (options: OfflineTransportOptions) => OfflineStore;
+
+export interface OfflineTransportOptions extends InternalBaseTransportOptions {
+  /**
+   * A function that creates the offline store instance.
+   */
+  createStore?: CreateOfflineStore;
+
+  /**
+   * Flush the offline store shortly after startup.
+   *
+   * Defaults: false
+   */
+  flushAtStartup?: boolean;
+
+  /**
+   * Called before an event is stored.
+   *
+   * Return false to drop the envelope rather than store it.
+   *
+   * @param envelope The envelope that failed to send.
+   * @param error The error that occurred.
+   * @param retryDelay The current retry delay in milliseconds.
+   */
+  shouldStore?: (envelope: Envelope, error: Error, retryDelay: number) => boolean | Promise<boolean>;
+}
+
+type Timer = number | { unref?: () => void };
+
+/**
+ * Wraps a transport and stores and retries events when they fail to send.
+ *
+ * @param createTransport The transport to wrap.
+ */
+export function makeOfflineTransport<TO>(
+  createTransport: (options: TO) => Transport,
+): (options: TO & OfflineTransportOptions) => Transport {
+  return options => {
+    const transport = createTransport(options);
+    const store = options.createStore ? options.createStore(options) : undefined;
+
+    let retryDelay = START_DELAY;
+    let flushTimer: Timer | undefined;
+
+    function shouldQueue(env: Envelope, error: Error, retryDelay: number): boolean | Promise<boolean> {
+      // We don't queue Session Replay envelopes because they are:
+      // - Ordered and Replay relies on the response status to know when they're successfully sent.
+      // - Likely to fill the queue quickly and block other events from being sent.
+      if (isReplayEnvelope(env)) {
+        return false;
+      }
+
+      if (options.shouldStore) {
+        return options.shouldStore(env, error, retryDelay);
+      }
+
+      return true;
+    }
+
+    function flushIn(delay: number): void {
+      if (!store) {
+        return;
+      }
+
+      if (flushTimer) {
+        clearTimeout(flushTimer as ReturnType<typeof setTimeout>);
+      }
+
+      flushTimer = setTimeout(async () => {
+        flushTimer = undefined;
+
+        const found = await store.pop();
+        if (found) {
+          log('Attempting to send previously queued event');
+          void send(found).catch(e => {
```

Comment:
Just to clarify: So if we try to re-send an envelope, we pop it from the store. but if it fails, we do not re-add it? So we would discard this envelope?

Reply:
No it will be re-added to the queued and we will continue to retry the envelope later. This means when offline, an envelope will be popped and re-inserted multiple times. The initial retry delay is only 5 seconds so if we don't retry, this is more of a "retry once" wrapper than offline.

I considered having a `peek` from the store and `delete` after successful send but this increases the chance of envelopes being sent multiple times, for example if `delete` fails or the page is closed between `peek`/`delete`.